### PR TITLE
add want-lgtm-all reusable workflow

### PR DESCRIPTION
**Pull request by verbanicm**
Porting the `want-lgtm-all` workflow from the hackathon. This changes from using labels to using a regex to find the `want_lgtm=all` string in the pull request body.

This also fixes a bug where submitted reviews could be request changes and not approved.

### CODE REVIEW

File: .github/workflows/want-lgtm-all.yml [hunk 1]
```diff
@@ -0,0 +1,165 @@
+name: 'want-lgtm-all'
+
+on:
+  pull_request:
+    types:
+      - opened
```
Comment (sethvargo) [inline .github/workflows/want-lgtm-all.yml:6]: Plz quote your YAML

File: .github/workflows/want-lgtm-all.yml [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+      - ready_for_review
+      - review_requested
+      - review_request_removed
+  pull_request_review:
+    types:
+      - submitted
+      - dismissed
+      - edited
+  workflow_call:
+  workflow_dispatch:
```
Comment (sethvargo) [inline .github/workflows/want-lgtm-all.yml:19]: Do we need both of these?

File: .github/workflows/want-lgtm-all.yml [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  validate:
+    runs-on: 'ubuntu-latest'
+    steps:
+      - id: 'validate-event'
+        name: 'Validate GitHub Event'
+        uses: 'actions/github-script@d7906e4ad0b1822421a7e6a35d5ca353c962f410' # ratchet:actions/github-script@v6
+        with:
+          script: |-
+            // validate required event names
+            if (!["pull_request", "pull_request_review"].includes(context.eventName)) {
```
Comment (sethvargo) [inline .github/workflows/want-lgtm-all.yml:39]: Can we do this as an `if` on the job instead?

File: .github/workflows/want-lgtm-all.yml [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+      - id: 'want-lgtm'
+        name: 'Validate PR Body'
+        uses: 'actions/github-script@d7906e4ad0b1822421a7e6a35d5ca353c962f410' # ratchet:actions/github-script@v6
+        with:
+          result-encoding: 'string'
+          script: |-
+            const body = context.payload.pull_request.body || "";
+
+            // if PR body does not contain want_lgtm_all, exit gracefully
```
Comment (sethvargo) [inline .github/workflows/want-lgtm-all.yml:54]: Nit: it's want_lgtm=(value) where "all" is a special value

File: README.md [hunk 5]
```diff
@@ -24,28 +22,27 @@ Use this workflow to perform basic Go linting checks:
 on:
   push:
     branches:
-      - 'main'
```
Comment (sethvargo) [inline README.md:27]: Why these changes? Also single-quoted YAML is safer.

File: .github/workflows/want-lgtm-all.yml [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            if (!body.match(/want_lgtm_all/gim)) {
```
Comment (sethvargo) [inline .github/workflows/want-lgtm-all.yml:55]: This seems far less efficient than `body.includes('want_lgtm_all')`? Even `body.toLowerCase().includes('want_lgtm_all')` would probably be more efficient.

**Review by sethvargo — COMMENTED** [review]
I mentioned it during the hackathon, but I'm worried about permissions here. This is fine for now, but it doesn't support a use case where a reviewer wants to add _another_ reviewer (or ask for another reviewer to LGTM something). That's a more difficult problem, because we don't want random people on the internet to be able to comment "want_lgtm=all" though.

File: README.md (hunk 5 above)
Comment (verbanicm) [reply README.md:27]: Errrr, my formatter must have done this for markdown.....I'll fix.

File: .github/workflows/want-lgtm-all.yml [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+      # status checks by github
+      - id: 'rerun-status-checks'
+        name: 'Re-run Status Checks'
+        if: ${{ steps.want-lgtm.outputs.result ==  'true' && github.event_name == 'pull_request_review' }}
+        uses: 'actions/github-script@d7906e4ad0b1822421a7e6a35d5ca353c962f410' # ratchet:actions/github-script@v6
+        with:
+          retries: 3
+          script: |-
+            // get the filename for this workflow
+            var workflowFilename = process.env.GITHUB_WORKFLOW_REF.split("@")[0]
```
Comment (volzmc) [inline .github/workflows/want-lgtm-all.yml:136]: nit: Prefer `let` or `const` over var

**Review by volzmc — APPROVED** [review]

File: .github/workflows/want-lgtm-all.yml [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+              workflow_id: workflowFilename,
+              branch: context.payload.pull_request.head.ref,
+              event: "pull_request",
+              status: "failure",
+              per_page: 100,
+            });
+
+            // filter workflow runs only for this pull request number
+            let unsuccessfulRuns = workflows.data.workflow_runs
+              .filter((w) =>
```
[the remark this answers is not in the document]
Comment (verbanicm) [reply .github/workflows/want-lgtm-all.yml:153]: I think we should write up a style guide... lol. I think ill leave it for now. But noted.

**Comment by verbanicm** [comment]
> I mentioned it during the hackathon, but I'm worried about permissions here. This is fine for now, but it doesn't support a use case where a reviewer wants to add _another_ reviewer (or ask for another reviewer to LGTM something). That's a more difficult problem, because we don't want random people on the internet to be able to comment "want_lgtm=all" though.

@sethvargo Added check to only allow owners, members and collaborators.

File: .github/workflows/want-lgtm-all.yml (hunk 3 above)
Comment (sethvargo) [reply .github/workflows/want-lgtm-all.yml]: Sorry, I'm still not seeing why we can't use a job `if`?

Comment (verbanicm) [reply .github/workflows/want-lgtm-all.yml]: I think you need to refresh.....

Comment (sethvargo) [reply .github/workflows/want-lgtm-all.yml]: You are correct.

**Review by sethvargo — APPROVED** [review]

File: test.json [hunk 9]
```diff
@@ -0,0 +1 @@
...,"html_url":"https://github.com/verbanicm","followers_url":"https://api.github.com/users/verbanicm/followers","following_url":"https://api.github.com/users/verbanicm/following{/other_user}","gists_url":"https://api.github.com/users/verbanicm/gists{/gist_id}","starred_url":"https://api.github.com/users/verbanicm/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/verbanicm/subscriptions","organizations_url":"https://api.github.com/users/verbanicm/orgs","repos_url":"https://api.github.com/users/verbanicm/repos","events_url":"https://api.github.com/users/verbanicm/events{/privacy}","received_events_url":"https://api.github.com/users/verbanicm/received_events","type":"User","site_admin":false},"body":"","state":"COMMENTED","html_url":"https://github.com/abcxyz/pkg/pull/116#pullrequestreview-1424904856","pull_request_url":"https://api.github.com/repos/abcxyz/pkg/pulls/116","author_association":"MEMBER","_links":{"html":{"href":"https://github.com/abcxyz/pkg/pull/116#pullrequestreview-1424904856"},"pull_request":{"href":"https://api.github.com/repos/abcxyz/pkg/pulls/116"}},"submitted_at":"2023-05-12T17:27:58Z","commit_id":"325c9e8751f11c896983fdbb1fcd3bbf53a3d040"}]
```
[the remark this answers is not in the document]
Comment (verbanicm) [reply test.json:1]: nah, slipped in there, ill take it out

**Review by sethvargo — APPROVED** [review]

**Review by bradegler — APPROVED** [review]

**Review by bradegler — APPROVED** [review]

**Review by bradegler — DISMISSED** [review]

**Review by bradegler — APPROVED** [review]
